### PR TITLE
Hugetlbfs support based on empty dir volume plugin

### PR DESCRIPTION
**Pull request by squall0gd**
**What this PR does / why we need it**: Support for huge pages in empty dir volume plugin. More information about hugepages can be found [here](https://www.kernel.org/doc/Documentation/vm/hugetlbpage.txt)

Feature track issue: kubernetes/features#275

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #

**Special notes for your reviewer**:

**Release note**:

```release-note
Support for Huge pages in empty_dir volume plugin
[Huge pages](https://www.kernel.org/doc/Documentation/vm/hugetlbpage.txt) can now be used with empty dir volume plugin.
```

### CODE REVIEW

File: pkg/volume/empty_dir/empty_dir.go
Comment (dixudx) [inline pkg/volume/empty_dir/empty_dir.go]: Does this mean that if we are using `StorageMediumHugetlbfs`, the `size` is also needed alongside?

If so, you'd better add some validations in `pkg/api/validation/validation.go`.

File: pkg/api/types.go
Comment (dixudx) [inline pkg/api/types.go]: It's kind of confusing to declare xxSize as `string`. May it be `int`?

File: pkg/volume/empty_dir/empty_dir.go
Comment (dixudx) [inline pkg/volume/empty_dir/empty_dir.go]: Same as above. Maybe `int`?

File: pkg/volume/empty_dir/empty_dir.go
Comment (squall0gd) [inline pkg/volume/empty_dir/empty_dir.go]: It's required in that case. I will move this validation to `pkg/api/validation/validation.go`

File: pkg/volume/empty_dir/empty_dir.go
Comment (squall0gd) [inline pkg/volume/empty_dir/empty_dir.go]: I think it's better to use `Quantity` here b/c I need to keep unit.

File: pkg/api/types.go
Comment (squall0gd) [inline pkg/api/types.go]: I'll use `Quantity` here :)

File: staging/src/k8s.io/api/core/v1/types.go
Comment (squall0gd) [inline staging/src/k8s.io/api/core/v1/types.go]: :+1:

**Comment by xiangpengzhao** [comment]
@squall0gd FYI, there is a proposal on HugePages here: https://github.com/kubernetes/community/pull/837

/cc @derekwaynecarr

**Comment by derekwaynecarr** [comment]
i want to make sure we have agreement on the proposal referenced before merging this, but thank you for your help!

File: pkg/api/types.go
Comment (derekwaynecarr) [inline pkg/api/types.go]: +1 to quantity

File: pkg/api/types.go
Comment (derekwaynecarr) [inline pkg/api/types.go]: prefer HugePages

File: pkg/volume/empty_dir/empty_dir.go
Comment (derekwaynecarr) [inline pkg/volume/empty_dir/empty_dir.go]: i don't understand why you have size.

size is bounded by sum resource requirement across containers on pod.

we need a new field specific for hugepages that aligns with pagesize. mount option.

Comment (squall0gd) [inline pkg/volume/empty_dir/empty_dir.go]: well I've used `size` w/o any reason - just want some, easy to track parameter, to verify that passing mount options is working correctly. Actually I'm waiting for proposal's merge to finalize implementing mount options and to play with unit tests. :)

File: pkg/volume/empty_dir/empty_dir.go
Comment (PiotrProkop) [inline pkg/volume/empty_dir/empty_dir.go]: Please fix this import.

File: pkg/volume/empty_dir/empty_dir.go
Comment (PiotrProkop) [inline pkg/volume/empty_dir/empty_dir.go]: delete this line

File: pkg/volume/empty_dir/empty_dir.go
Comment (PiotrProkop) [inline pkg/volume/empty_dir/empty_dir.go]: Please add a comment why mountOptions are empty.

File: pkg/volume/empty_dir/empty_dir.go
Comment (rootfs) [inline pkg/volume/empty_dir/empty_dir.go]: remove blank line

**Comment by rootfs** [comment]
can you add unit test cases too?

**Comment by derekwaynecarr** [comment]
I would like https://github.com/kubernetes/kubernetes/pull/50859 to merge before this.

File: examples/volumes/empty_dir/README.md
Comment (derekwaynecarr) [inline examples/volumes/empty_dir/README.md]: we can provide an example in a follow-on, split this into a separate pr that is a follow-on.

File: pkg/api/types.go
Comment (derekwaynecarr) [inline pkg/api/types.go]: s/Hugepages/HugePages per https://github.com/kubernetes/community/blob/master/contributors/design-proposals/hugepages.md#pod-specification

File: pkg/volume/empty_dir/empty_dir.go
Comment (derekwaynecarr) [inline pkg/volume/empty_dir/empty_dir.go]: rebase against https://github.com/kubernetes/kubernetes/pull/50859 as it looks like its stabilized.

**Review by derekwaynecarr — CHANGES_REQUESTED** [review]
please rebase against https://github.com/kubernetes/kubernetes/pull/50859

as part of that rebase, ensure that usage of the HugePages medium in an EmptyDir is restricted by the feature gate flag during pod validation so we can prevent usage of the medium to only the feature being enabled.

**Comment by PiotrProkop** [comment]
I cherry-picked commits from #50859 and applied @derekwaynecarr comments.

File: pkg/volume/empty_dir/empty_dir.go
Comment (derekwaynecarr) [reply pkg/volume/empty_dir/empty_dir.go]: nit: this was missed.

**Comment by derekwaynecarr** [comment]
just the one nit that can be done in follow-on, otherwise this is good for me.

will want someone from storage sig to look.

/cc @childsb @eparis

**Review by derekwaynecarr — APPROVED** [review]

**Comment by PiotrProkop** [comment]
Thanks for review @derekwaynecarr , I have just deleted that const.

File: pkg/volume/empty_dir/empty_dir.go
Comment (wongma7) [inline pkg/volume/empty_dir/empty_dir.go]: "medium hugepages"

**Comment by wongma7** [comment]
emptydir changes lgtm, jsut comment nit

File: pkg/volume/empty_dir/empty_dir.go
Comment (PiotrProkop) [reply pkg/volume/empty_dir/empty_dir.go]: OK fixing it :D

**Comment by derekwaynecarr** [comment]
/lgtm

Thanks @squall0gd for your help getting this support.

**Comment by jdumars** [comment]
Curious how this got merged despite failing cross:
I0905 18:14:23.457] pkg/volume/empty_dir/empty_dir_linux.go:54: constant 2508478710 overflows int32
@squall0gd did this error always present in testing and it just got missed?

**Comment by squall0gd** [comment]
@jdumars didn't notice this error before. Just some timeouts on e2e gates

**Comment by jdumars** [comment]
The test will need to be updated in order to accommodate the change.

**Comment by squall0gd** [comment]
@jdumars Well it looks like we need to add type to magic number constant. Sadly we need to do this in another PR.

**Comment by liggitt** [comment]
opened https://github.com/kubernetes/kubernetes/pull/51984 to type-cast the comparison to the constant

**Pull request merged**
